### PR TITLE
[ValueTracking] Respect `samesign` flag in `isKnownInversion`

### LINKED ISSUE #112350: [InstCombine] `samesign` flag should be dropped
Reproducer:
```
define i1 @src(i64 %x, i64 %y, i1 %cond) {
  %cmp1 = icmp ult i64 %x, %y
  %cmp2 = icmp samesign uge i64 %x, %y
  %sel = select i1 %cond, i1 %cmp1, i1 %cmp2
  ret i1 %sel
}

define i1 @tgt(i64 %x, i64 %y, i1 %cond) {
  %cmp2 = icmp samesign uge i64 %x, %y
  %sel = xor i1 %cond, %cmp2
  ret i1 %sel
}
```

### PR DESCRIPTION
**Pull request by dtcxzyw**
In https://github.com/llvm/llvm-project/pull/93591 we introduced `isKnownInversion` and assumes `X` is poison implies `Y` is poison because they share common operands. But after introducing `samesign` this assumption no longer hold if `X` is an icmp has `samesign` flag.

Alive2 link: https://alive2.llvm.org/ce/z/rj3EwQ (Please run it locally with this patch and https://github.com/AliveToolkit/alive2/pull/1098).

This approach is the most conservative way in my mind to address this problem. If `X` has `samesign` flag, it will check if `Y` also has this flag and make sure constant RHS operands have the same sign.

Pros: No API breaking/fast
Cons: less optimization opportunities (But I believe such cases don't exist in real-world code)

I also tried the following solutions:

+ Alternative solution 1: Check `impliesPoison` explicitly and insert `freeze` on demand
    + Pros: instruction-independent
    + Cons: increased compilation time/additional freeze nodes/break API semantic

+ Alternative solution 2: Add an optional `DropFlags` parameter and drop flags in InstCombine
    + Pros: instruction-independent/more optimization opportunities
    + Cons: increased compilation time due to re-enqueue/overhead with small-vector/API breaking

+ Alternative solution 3: Drop all poison-generating flags in InstCombine if `FalseVal` is an ICmp/Instruction
    + Pros: simple/fast
    + Cons: instruction-dependent/hard to extend

Fixes https://github.com/llvm/llvm-project/issues/112350.

### CODE REVIEW

**Review by antoniofrighetto — COMMENTED** [review]
> * Alternative solution 3: Drop all poison-generating flags in InstCombine if `FalseVal` is an ICmp/Instruction
>   
>   * Pros: simple/fast
>   * Cons: instruction-dependent/hard to extend

Out of curiosity, how come is it a cons if it happens to be instruction-dependant? It doesn't seem to be necessarily so to me? Current solution looks fine to me.

**Comment by dtcxzyw** [comment]
> Out of curiosity, how come is it a cons if it happens to be instruction-dependant? It doesn't seem to be necessarily so to me?

If we handle some complex patterns in `isKnownInversion` (i.e., we should drop flags in instructions other than `FalseVal` to ensure poison-safety), this approach doesn't work.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -8526,14 +8526,26 @@ bool llvm::isKnownInversion(const Value *X, const Value *Y) {
       !match(Y, m_c_ICmp(Pred2, m_Specific(A), m_Value(C))))
     return false;
 
+  // If they are only differ in predicate. They must both have samesign flag or
+  // not.
   if (B == C)
-    return Pred1 == ICmpInst::getInversePredicate(Pred2);
+    return Pred1 == ICmpInst::getInversePredicate(Pred2) &&
+           (!cast<ICmpInst>(X)->hasSameSign() ||
+            cast<ICmpInst>(Y)->hasSameSign());
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:8534]: Agree.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**